server/aesdsocket.c

```diff
@@ -0,0 +1,380 @@
+    int file_bytes = 0;                         // Number of bytes present in fd
+
+    while(1)
+    {
+        // Connection with the client
+        client_fd = accept(sfd, (struct sockaddr *)&test_addr, &addr_size);
+        if(client_fd == -1)
+        {
+            if(storage_buffer)
+            {
+                free(storage_buffer);
+                storage_buffer = NULL;
+            }
+            perror("accept failure: ");
+            continue;                           // Try again
+        }
+
+        if(complete_flag)
+        {
+            /* Buffer for receiving data from client */
+            storage_buffer = (char *)malloc(INIT_BUF_SIZE);
+            if(!storage_buffer)
+            {
+                perror("malloc failure: ");
+                return -1;
+            }
+            complete_flag = 0;
+        }
+
+        // Print the IP address of the client.
+        struct sockaddr_in *p = (struct sockaddr_in *)&test_addr;
+        syslog(LOG_DEBUG, "Accepted connection from %s", inet_ntop(AF_INET, &p->sin_addr, address_string, sizeof(address_string)));
+
+        // Receive bytes by polling the client fd
+        while((bytes_rcvd = recv(client_fd, buffer, sizeof(buffer), 0)) > 0)
+        {
+            printf("bytes rcvd = %d\n", bytes_rcvd);
+
+            for(int i=0; i < bytes_rcvd; i++)
+            {
+                if(buffer[i] == '\n')
+                {
+                    complete_flag = 1;
+                    actual_bytes_rcvd = i+1;
+                    break;
+                }
+            }
+            memcpy(storage_buffer + (storage_buffer_size_cnt - 1)*INIT_BUF_SIZE, buffer, bytes_rcvd);
+
+            if(complete_flag == 1)
+            {
+                bytes_to_write = (storage_buffer_size_cnt - 1)*INIT_BUF_SIZE + actual_bytes_rcvd;
+                break;
+            }
+            else    // reallocated the buffer by multiples of 1024 until complete flag is set
+            {
+                char *temp_ptr = (char *)realloc(storage_buffer, (storage_buffer_size + INIT_BUF_SIZE));
+                if(temp_ptr)
+                {
+                    storage_buffer = temp_ptr;
+                    storage_buffer_size += INIT_BUF_SIZE;
+                    storage_buffer_size_cnt += 1;
+                }
+                else
+                {
+                    printf("No memory available to alloc\n");
+                }
+            }
+        }
+        printf("Storage buffer size = %d\n", bytes_to_write);
+
+        // If transmission is complete,
+        // Copy to local fd and then rewrite/resend it back to client.
+        if(1 == complete_flag)
+        {
+            //Write to file
+            int wr_bytes = write(fd, storage_buffer, bytes_to_write);
+            file_bytes += wr_bytes;
+
+            //Set the fd to start of the file.
+            lseek(fd, 0, SEEK_SET);
+
+            read_buffer = (char *)malloc(file_bytes);
```

Comment:
@dash6424 what assumption are you making here about the size of the heap relative to the size of all packets sent?